### ISSUE TITLE
failed to allocate peer tty device

### ISSUE THREAD

**Issue by overlayfs** [lxc/lxc#1552]
**Issue**
lxc-attach fails with error message _failed to allocate peer tty device_ in certain situations.  This issue only arose within the last few days and may be caused by the recent commit _console: fail when we cannot allocate peer tty_.

**Steps to reproduce**
- `lxc-create -n test -t download -- -d ubuntu -r trusty -a amd64`  # unprivileged
- `lxc-start -n test`
- Use the _run command_ dialogue (**Alt+F2**) to execute the following:
   `lxc-attach -n test -o /tmp/lxc-attach.log -l DEBUG touch /tmp/lxc-test`

**Error message**
lxc-attach test xxx5.342 INFO     lxc_attach_ui - tools/lxc_attach.c:get_pty_on_host:288 - Trying to allocate a pty on the host
lxc-attach test xxx5.343 ERROR    console - console.c:lxc_console_peer_default:444 - failed to open "/dev/tty"
lxc-attach test xxx5.343 ERROR    console - console.c:lxc_console_create:549 - failed to allocate peer tty device

**OS details**
-  lxc-daily ppa
- host & container: ubuntu-14.04 64bit
- kernel: linux-generic-lts-xenial

**Comment by brauner** [comment]
Can please start the container with debug enabled and also give me the output of the lxc-attach debug log, please.

**Comment by brauner** [comment]
Hm, I know what this is caused by. But I need to think about possible security implications.

**Comment by overlayfs** [comment]
**lxc-start.log**

      lxc-start test 20170512134254.262 DEBUG    lxc_monitor - monitor.c:lxc_monitord_spawn:330 - Going to wait for pid 24659.
      lxc-start test 20170512134254.263 DEBUG    lxc_monitor - monitor.c:lxc_monitord_spawn:349 - Trying to sync with child process.
      lxc-start test 20170512134254.263 INFO     lxc_start - start.c:lxc_check_inherited:243 - Closed inherited fd: 3.
      lxc-start test 20170512134254.263 INFO     lxc_start - start.c:lxc_check_inherited:243 - Closed inherited fd: 5.
      lxc-start test 20170512134254.263 DEBUG    lxc_monitor - monitor.c:lxc_monitord_spawn:387 - Using pipe file descriptor 6 for monitord.
      lxc-start test 20170512134254.273 DEBUG    lxc_monitor - monitor.c:lxc_monitord_spawn:364 - Successfully synced with child process.
      lxc-start test 20170512134254.274 DEBUG    lxc_monitor - monitor.c:lxc_monitord_spawn:333 - Finished waiting on pid 24659.
      lxc-start test 20170512134254.275 INFO     lxc_container - lxccontainer.c:do_lxcapi_start:802 - Attempting to set proc title to [lxc monitor] /home/oleg/.local/share/lxc junk
      lxc-start test 20170512134254.277 DEBUG    lxc_monitor - monitor.c:lxc_monitord_spawn:330 - Going to wait for pid 24663.
      lxc-start test 20170512134254.277 INFO     lxc_start - start.c:lxc_check_inherited:243 - Closed inherited fd: 3.
      lxc-start test 20170512134254.277 INFO     lxc_lsm - lsm/lsm.c:lsm_init:48 - LSM security driver AppArmor
      lxc-start test 20170512134254.277 INFO     lxc_seccomp - seccomp.c:parse_config_v2:418 - processing: .reject_force_umount  # comment this to allow umount -f;  not recommended.
      lxc-start test 20170512134254.277 DEBUG    lxc_monitor - monitor.c:lxc_monitord_spawn:349 - Trying to sync with child process.
      lxc-start test 20170512134254.277 INFO     lxc_seccomp - seccomp.c:parse_config_v2:590 - Adding native rule for reject_force_umount action 0.
      lxc-start test 20170512134254.277 INFO     lxc_seccomp - seccomp.c:do_resolve_add_rule:259 - Setting Seccomp rule to reject force umounts.
      lxc-start test 20170512134254.277 INFO     lxc_seccomp - seccomp.c:parse_config_v2:593 - Adding compat rule for reject_force_umount action 0.
      lxc-start test 20170512134254.277 INFO     lxc_seccomp - seccomp.c:do_resolve_add_rule:259 - Setting Seccomp rule to reject force umounts.
      lxc-start test 20170512134254.277 INFO     lxc_seccomp - seccomp.c:do_resolve_add_rule:259 - Setting Seccomp rule to reject force umounts.
      lxc-start test 20170512134254.277 INFO     lxc_seccomp - seccomp.c:parse_config_v2:418 - processing: .[all].
      lxc-start test 20170512134254.277 INFO     lxc_seccomp - seccomp.c:parse_config_v2:418 - processing: .kexec_load errno 1.
      lxc-start test 20170512134254.277 INFO     lxc_seccomp - seccomp.c:parse_config_v2:590 - Adding native rule for kexec_load action 327681.
      lxc-start test 20170512134254.277 INFO     lxc_seccomp - seccomp.c:parse_config_v2:593 - Adding compat rule for kexec_load action 327681.
      lxc-start test 20170512134254.277 INFO     lxc_seccomp - seccomp.c:parse_config_v2:418 - processing: .open_by_handle_at errno 1.
      lxc-start test 20170512134254.277 INFO     lxc_seccomp - seccomp.c:parse_config_v2:590 - Adding native rule for open_by_h...skipping...
      lxc-start test 20170512134404.513 DEBUG    lxc_commands - commands.c:lxc_cmd_handler:901 - Peer has disconnected.
         lxc-ls 20170512134404.514 DEBUG    lxc_commands - commands.c:lxc_cmd_rsp_recv:203 - command get_init_pid response data length is 0
      lxc-start test 20170512134404.514 DEBUG    lxc_commands - commands.c:lxc_cmd_handler:901 - Peer has disconnected.
         lxc-ls 20170512134404.515 DEBUG    lxc_commands - commands.c:lxc_cmd_rsp_recv:203 - command get_init_pid response data length is 0
      lxc-start test 20170512134404.515 DEBUG    lxc_commands - commands.c:lxc_cmd_handler:901 - Peer has disconnected.
         lxc-ls 20170512134404.517 DEBUG    lxc_commands - commands.c:lxc_cmd_rsp_recv:203 - command get_init_pid response data length is 0
      lxc-start test 20170512134404.517 DEBUG    lxc_commands - commands.c:lxc_cmd_handler:901 - Peer has disconnected.
      lxc-start test 20170512134404.518 DEBUG    lxc_commands - commands.c:lxc_cmd_handler:901 - Peer has disconnected.
         lxc-ls 20170512134404.518 DEBUG    lxc_commands - commands.c:lxc_cmd_rsp_recv:203 - command get_state response data length is 0
         lxc-ls 20170512134404.518 DEBUG    lxc_commands - commands.c:lxc_cmd_get_state:586 - Container "test" is in "RUNNING" state.
      lxc-start test 20170512134404.518 DEBUG    lxc_commands - commands.c:lxc_cmd_handler:901 - Peer has disconnected.
      lxc-start test 20170512134404.518 DEBUG    lxc_commands - commands.c:lxc_cmd_handler:901 - Peer has disconnected.
      lxc-start test 20170512134404.518 DEBUG    lxc_commands - commands.c:lxc_cmd_handler:901 - Peer has disconnected.
         lxc-ls 20170512134404.518 DEBUG    lxc_commands - commands.c:lxc_cmd_rsp_recv:203 - command get_state response data length is 0
         lxc-ls 20170512134404.518 DEBUG    lxc_commands - commands.c:lxc_cmd_get_state:586 - Container "test" is in "RUNNING" state.
      lxc-start test 20170512134404.518 DEBUG    lxc_commands - commands.c:lxc_cmd_handler:901 - Peer has disconnected.
      lxc-start test 20170512134404.518 DEBUG    lxc_commands - commands.c:lxc_cmd_handler:901 - Peer has disconnected.
      lxc-start test 20170512134404.518 DEBUG    lxc_commands - commands.c:lxc_cmd_handler:901 - Peer has disconnected.
         lxc-ls 20170512134404.518 DEBUG    lxc_commands - commands.c:lxc_cmd_rsp_recv:203 - command get_state response data length is 0
         lxc-ls 20170512134404.519 DEBUG    lxc_commands - commands.c:lxc_cmd_get_state:586 - Container "test" is in "RUNNING" state.
      lxc-start test 20170512134404.519 DEBUG    lxc_commands - commands.c:lxc_cmd_handler:901 - Peer has disconnected.
      lxc-start test 20170512134404.519 DEBUG    lxc_commands - commands.c:lxc_cmd_handler:901 - Peer has disconnected.
      lxc-start test 20170512134404.519 DEBUG    lxc_commands - commands.c:lxc_cmd_handler:901 - Peer has disconnected.
         lxc-ls 20170512134404.519 DEBUG    lxc_commands - commands.c:lxc_cmd_rsp_recv:203 - command get_state response data length is 0
         lxc-ls 20170512134404.519 DEBUG    lxc_commands - commands.c:lxc_cmd_get_state:586 - Container "test" is in "RUNNING" state.
      lxc-start test 20170512134404.519 DEBUG    lxc_commands - commands.c:lxc_cmd_handler:901 - Peer has disconnected.
      lxc-start test 20170512134404.519 DEBUG    lxc_commands - commands.c:lxc_cmd_handler:901 - Peer has disconnected.
      lxc-start test 20170512134420.399 DEBUG    lxc_commands - commands.c:lxc_cmd_handler:901 - Peer has disconnected.

**Comment by overlayfs** [comment]
**lxc-attach.log**

     lxc-attach test 20170512134420.398 WARN     lxc_confile - confile.c:config_pivotdir:2104 - lxc.pivotdir is ignored.  It will soon become an error.
     lxc-attach test 20170512134420.399 INFO     lxc_confile - confile.c:config_idmap:1725 - read uid map: type u nsid 0 hostid 100000 range 65536
     lxc-attach test 20170512134420.399 INFO     lxc_confile - confile.c:config_idmap:1725 - read uid map: type g nsid 0 hostid 100000 range 65536
     lxc-attach test 20170512134420.399 DEBUG    lxc_commands - commands.c:lxc_cmd_rsp_recv:203 - command get_init_pid response data length is 0
     lxc-attach test 20170512134420.399 INFO     lxc_attach_ui - tools/lxc_attach.c:get_pty_on_host:288 - Trying to allocate a pty on the host
     lxc-attach test 20170512134420.400 ERROR    console - console.c:lxc_console_peer_default:444 - failed to open "/dev/tty"
     lxc-attach test 20170512134420.401 ERROR    console - console.c:lxc_console_create:549 - failed to allocate peer tty device

**Comment by brauner** [comment]
Yeah, you're getting `ENXIO` because `/dev/tty` doesn't exist. Sending a small fix to `lxc_attach.c`.

**Comment by overlayfs** [comment]
Commit 6f18b9c does indeed fix the issue.

Thank you Christian!

**Comment by gstanden** [comment]
Hi I am running into a situation very similar and I think this started happening in lxc 2.0.8 (I was not getting this issue in lxc 2.0.5).  When I start a container on openvswitch, it starts ok THE FIRST TIME ONLY.  Also, after the first startup, lxc-attach does not work.  I get this message:  

ubuntu@zesty:~/Downloads/orabuntu-lxc-master/orabuntu$ sudo lxc-attach -n oel73
lxc-attach: console.c: lxc_console_create: 533 Invalid argument - failed to allocate a pty

When I shutdown the container and try to restart it, it fails to start and gives this message, and moreover, prevents all other containers from starting from that point onward as well.

ubuntu@zesty:~/Downloads/orabuntu-lxc-master/orabuntu$ sudo lxc-start -n oel73 -F
lxc-start: console.c: lxc_console_create: 533 Invalid argument - failed to allocate a pty
lxc-start: start.c: lxc_init: 466 Failed to create console for container "oel73".
lxc-start: start.c: __lxc_start: 1314 Failed to initialize container "oel73".
lxc-start: tools/lxc_start.c: main: 366 The container failed to start.
lxc-start: tools/lxc_start.c: main: 370 Additional information can be obtained by setting the --logfile and --logpriority options.

This is really causing a big problem, and I wasn't seeing this until lxc 2.0.8 rolled out.  I'm using the lxc from the ubuntu package repo installed from apt-get.  Do you have some way for me to fix this?

Thanks

**Comment by brauner** [comment]
@gstanden, can you start the container with `lxc-start -n <container-name> -l trace -o <container-name>-log` and append the log file or paste it's output here, please?

**Comment by gstanden** [comment]
Hi Christian thank you very much for replying.  

So when the containers are first built, they are on the standard lxcbr0 that comes with the Ubuntu package version of lxc (currently 2.0.8).  Everything works great when they are on lxcbr0

So I'm going to provide what you asked plus some additional information.  I hope it helps and is not extraneous.

During the deployment, a container called 'olive' is created on lxcbr0 and then moved to openvswitch.  At this stage, olive starts and stops normally with no issues, while on openvswitch, and lxc-attach works normally with no issues:  I can stop/start/attach with no issues.  At this phase, the setup is olive is on openvswitch, and oel73 is still on lxcbr0:

### normal no problems at this point ###
(oel73 on lxcbr0, and olive on openvswitch OVS)

ubuntu@skynet3:~$ sudo lxc-ls -f
[sudo] password for ubuntu: 
NAME  STATE   AUTOSTART GROUPS IPV4                     IPV6 
oel73 STOPPED 0         -      -                        -    
olive RUNNING 0           -      10.207.29.2, 10.207.39.2 -    
ubuntu@skynet3:~$ sudo lxc-start -n oel73
ubuntu@skynet3:~$ sudo lxc-attach -n oel73
bash-4.2# exit
exit
ubuntu@skynet3:~$ sudo lxc-ls -f
NAME  STATE   AUTOSTART GROUPS IPV4                     IPV6 
oel73 RUNNING 0         -      10.0.3.226    (lxcbr0)           -    
olive RUNNING 0          -      10.207.29.2, 10.207.39.2 -    (openvswitch)
ubuntu@skynet3:~$ sudo lxc-attach -n oel73
bash-4.2# exit
exit
ubuntu@skynet3:~$ sudo lxc-attach -n olive
root@olive:~# exit
exit
ubuntu@skynet3:~$ sudo lxc-stop -n oel73
ubuntu@skynet3:~$ sudo lxc-stop -n olive
ubuntu@skynet3:~$ sudo lxc-ls -f
NAME  STATE   AUTOSTART GROUPS IPV4 IPV6 
oel73 STOPPED 0         -      -    -    
olive  STOPPED 0         -      -    -    
ubuntu@skynet3:~$ sudo lxc-start -n oel73 -l trace -o oel73.txt
ubuntu@skynet3:~$ sudo lxc-start -n olive -l trace -o olive.txt
ubuntu@skynet3:~$ sudo lxc-ls -f
NAME  STATE   AUTOSTART GROUPS IPV4                     IPV6 
oel73 RUNNING 0         -      10.0.3.226               -    
olive  RUNNING 0         -      10.207.29.2, 10.207.39.2 -    
ubuntu@skynet3:~$ sudo lxc-attach -n olive
root@olive:~# exit
exit
ubuntu@skynet3:~$ sudo lxc-attach -n oel73
bash-4.2# exit
exit
ubuntu@skynet3:~$

### normal no problems at this point ###

Here are the config files at this point:

## oel73 config ###
root@skynet3:/var/lib/lxc/oel73# cat config
# Template used to create this container: /usr/share/lxc/templates/lxc-oracle
# Parameters passed to the template: --release=7.3
# Template script checksum (SHA-1): fa97092e4ad32594f23d228c57d78246a9e1c15b
# For additional config options, please look at lxc.container.conf(5)

# Uncomment the following line to support nesting containers:
#lxc.include = /usr/share/lxc/config/nesting.conf
# (Be aware this has security implications)

lxc.network.type = veth
lxc.network.link = lxcbr0
lxc.network.flags = up
lxc.network.hwaddr = 00:16:3e:3d:75:2a
lxc.rootfs = /var/lib/lxc/oel73/rootfs
lxc.rootfs.backend = dir
# Common configuration
lxc.include = /usr/share/lxc/config/oracle.common.conf
# Container configuration for Oracle Linux 7.3
lxc.arch = x86_64
lxc.utsname = oel73
lxc.cap.drop = sys_resource
# Networking
lxc.network.name = eth0
lxc.network.mtu = 1500
root@skynet3:/var/lib/lxc/oel73#

### olive config ###
root@skynet3:/var/lib/lxc/olive# cat config
# Template used to create this container: /usr/share/lxc/templates/lxc-ubuntu
# Parameters passed to the template: -r xenial -a amd64
# For additional config options, please look at lxc.container.conf(5)

# Uncomment the following line to support nesting containers:
#lxc.include = /usr/share/lxc/config/nesting.conf
# (Be aware this has security implications)


# Common configuration
lxc.include = /usr/share/lxc/config/ubuntu.common.conf

# Container specific configuration
lxc.rootfs = /var/lib/lxc/olive/rootfs
lxc.rootfs.backend = dir
lxc.utsname = olive
lxc.arch = amd64

# OpenvSwitch Networking
lxc.network.type = veth
lxc.network.flags = up
lxc.network.script.up = /etc/network/if-up.d/openvswitch/olive-pub-ifup-sw1
lxc.network.script.down = /etc/network/if-down.d/openvswitch/olive-pub-ifdown-sw1
lxc.network.veth.pair = olivew
lxc.network.name = eth0
lxc.network.mtu = 1500
lxc.network.hwaddr = 00:16:3e:ce:de:25
# lxc.network.ipv4 = 10.207.39.2/24

lxc.network.type = veth
lxc.network.flags = up
lxc.network.script.up = /etc/network/if-up.d/openvswitch/olive-pub-ifup-sx1
lxc.network.script.down = /etc/network/if-down.d/openvswitch/olive-pub-ifdown-sx1
lxc.network.veth.pair = olivex
lxc.network.name = eth1
lxc.network.mtu = 1500
lxc.network.hwaddr = 00:16:3e:ce:de:26
# lxc.network.ipv4 = 10.207.29.2/24
root@skynet3:/var/lib/lxc/olive# 

I've been using this config for 'olive' for a long time, it has always worked with no issues.  Here are the script.up and script.down:

root@skynet3:/var/lib/lxc/olive# cat /etc/network/if-up.d/openvswitch/olive-pub-ifup-sw1
#!/bin/bash
ovsBr='sw1'
ovs-vsctl add-port ${ovsBr} $5
ovs-vsctl set port $5 tag=10
root@skynet3:/var/lib/lxc/olive# cat /etc/network/if-down.d/openvswitch/olive-pub-ifdown-sw1
#!/bin/bash
ovsBr='sw1'
sudo ovs-vsctl del-port ${ovsBr} $5
function CheckPortExists
{
sudo ovs-vsctl show | grep \"$5\" | grep Port | sed 's/^[ \t]*//;s/[ \t]*$//' | cut -f2 -d' ' | sed 's/"//g'
}
PortExists=$(CheckPortExists)
if [ "$PortExists" = "$5" ]
then
sudo ovs-vsctl del-port ${ovsBr} $5
fi
root@skynet3:/var/lib/lxc/olive# 

I've attached the logs you requested for this configuration where all is working correctly as olive.log and oel73.log.  These are the logs for the no-issues everything working fine case.

[olive.txt](https://github.com/lxc/lxc/files/1129738/olive.txt)
[oel73.txt](https://github.com/lxc/lxc/files/1129735/oel73.txt)

### End of configuration that is working fine.

### New configuration - both containers on openvswitch now ###

Ok, so now I move oel73 over to openvswitch also so that both containers are on openvswitch.  of course olive config stays the same, but oel73 config is changed to this:

ubuntu@skynet3:~$ sudo lxc-ls -f
NAME  STATE   AUTOSTART GROUPS IPV4                     IPV6 
oel73 RUNNING 0         -      10.0.3.226               -    
olive RUNNING 0         -      10.207.29.2, 10.207.39.2 -    
ubuntu@skynet3:~$ sudo lxc-stop -n oel73

ubuntu@skynet3:~$ sudo su -

root@skynet3:~# cd /var/lib/lxc/oel73
root@skynet3:/var/lib/lxc/oel73# cp -p config config.original.bak
root@skynet3:/var/lib/lxc/oel73# cp -p config.oracle.bak.oel7 config
root@skynet3:/var/lib/lxc/oel73# vi config
root@skynet3:/var/lib/lxc/oel73# cat config
# Template used to create this container: /usr/share/lxc/templates/lxc-oracle
# Parameters passed to the template: --release=7.1
# For additional config options, please look at lxc.container.conf(5)
lxc.rootfs = /var/lib/lxc/oel73/rootfs
# Common configuration
lxc.include = /usr/share/lxc/config/oracle.common.conf
# Container configuration for Oracle Linux 7.1
lxc.arch = x86_64
lxc.utsname = oel73
lxc.cap.drop = sys_resource
# Networking
lxc.network.type = veth
lxc.network.flags = up
lxc.network.script.up = /etc/network/if-up.d/openvswitch/oel73-pub-ifup-sx1
lxc.network.script.down = /etc/network/if-down.d/openvswitch/oel73-pub-ifdown-sx1
lxc.network.veth.pair = oel73
lxc.network.name = eth0
lxc.network.mtu = 1500
lxc.network.hwaddr = fe:88:ac:95:c3:98
# Devices, Filesystems, Mounts
lxc.cgroup.devices.allow = a # Allow access to all devices
lxc.cgroup.devices.allow = a 10:236 rwm
lxc.cgroup.devices.allow = a 252:* rwm
lxc.cgroup.devices.allow = a 8:* rwm
lxc.mount.entry = /dev/mapper /var/lib/lxc/oel73/rootfs/dev/mapper none defaults,bind,create=dir 0 0
lxc.mount.entry = /dev /var/lib/lxc/oel73/rootfs/dev none defaults,bind,create=dir 0 0
lxc.mount.entry = shm dev/shm tmpfs size=3500m,nosuid,nodev,noexec,create=dir 0 0
lxc.mount.auto = proc sys cgroup
lxc.mount.auto = proc:rw sys:rw cgroup-full:rw
lxc.aa_allow_incomplete = 1
# Autostart at boot
# lxc.start.auto = 1
# lxc.group = onboot
# GLS 2014-10-23 NFS
lxc.aa_profile = unconfined

## Here are the lxc.network.script.up and lxc.network.script.down.  Note that they use VLAN tag 11 and the bridge is openvswitch sx1.

root@skynet3:/etc/network/if-down.d/openvswitch# cat /etc/network/if-up.d/openvswitch/oel73-pub-ifup-sx1
#!/bin/bash
ovsBr='sx1'
ovs-vsctl add-port ${ovsBr} $5
ovs-vsctl set port $5 tag=11

root@skynet3:/etc/network/if-down.d/openvswitch# cat /etc/network/if-down.d/openvswitch/oel73-pub-ifdown-sx1
#!/bin/bash
ovsBr='sx1'
sudo ovs-vsctl del-port ${ovsBr} $5
function CheckPortExists
{
sudo ovs-vsctl show | grep \"$5\" | grep Port | sed 's/^[ \t]*//;s/[ \t]*$//' | cut -f2 -d' ' | sed 's/"//g'
}
PortExists=$(CheckPortExists)
if [ "$PortExists" = "$5" ]
then
sudo ovs-vsctl del-port ${ovsBr} $5
fi
sudo ip link del $5 

## For reference here is the openvswitch bridge as well.  The bridge sx1 has VLAN 11.

root@skynet3:/etc/network/if-down.d/openvswitch# ovs-vsctl show
2ab10e06-086c-4a4b-a614-1492c0cceb56
    Bridge "sx1"
        Port "a1"
            tag: 11
            Interface "a1"
        Port "sx1"
            tag: 11
            trunks: [11]
            Interface "sx1"
                type: internal
        Port olivex
            tag: 11
            Interface olivex
    Bridge "sw1"
        Port "s2"
            tag: 10
            Interface "s2"
        Port olivew
            tag: 10
            Interface olivew
        Port "sw1"
            tag: 10
            trunks: [10]
            Interface "sw1"
                type: internal
        Port "s3"
            tag: 10
            Interface "s3"
        Port "s1"
            tag: 10
            Interface "s1"
    ovs_version: "2.6.1"
root@skynet3:/etc/network/if-down.d/openvswitch#

## Now I start up oel73 and it starts up normally (it gets its ip from dhcp isc-dhcp-server).  Note that oel73 only starts up successfully the first time after being moved to openvswitch with the new config file.  It won't startup anymore after it is shut down after this first startup (which starts up ok the first time only). Below is the logfile run for the "first start" of oel73 (and it does startup the first time only after being moved to openvswitch, but never again after being shutdown again).

ubuntu@skynet3:~$ sudo lxc-ls -f
NAME  STATE   AUTOSTART GROUPS IPV4                     IPV6 
oel73 STOPPED 0         -      -                        -    
olive RUNNING 0         -      10.207.29.2, 10.207.39.2 -    
ubuntu@skynet3:~$ sudo lxc-start -n oel73 -l trace -o oel73-on-ovs-first-start.txt
ubuntu@skynet3:~$ sudo lxc-ls -f
NAME  STATE   AUTOSTART GROUPS IPV4                     IPV6 
oel73 RUNNING 0         -      -                        -    
olive RUNNING 0         -      10.207.29.2, 10.207.39.2 -    
ubuntu@skynet3:~$ sudo lxc-ls -f
NAME  STATE   AUTOSTART GROUPS IPV4                     IPV6 
oel73 RUNNING 0         -      10.207.29.10             -    
olive RUNNING 0         -      10.207.29.2, 10.207.39.2 -    

[oel73-on-ovs-first-start.txt](https://github.com/lxc/lxc/files/1129763/oel73-on-ovs-first-start.txt)

But now lxc-attach no longer works (see below)  and it stops working for BOTH containers oel73 and olive (apparently because of something oel73 startup caused because up to this point lxc-attach to olive worked fine but as soon as oel73 started up it also breaks lxc-attach for olive too).

ubuntu@skynet3:~$ sudo lxc-attach -n oel73
lxc-attach: console.c: lxc_console_create: 533 Invalid argument - failed to allocate a pty
ubuntu@skynet3:~$ sudo lxc-attach -n olive
lxc-attach: console.c: lxc_console_create: 533 Invalid argument - failed to allocate a pty
ubuntu@skynet3:~$ 

### So naturally I stop both containers and try to restart them, but no-go:

ubuntu@skynet3:~$ sudo lxc-ls -f
NAME  STATE   AUTOSTART GROUPS IPV4                     IPV6 
oel73 RUNNING 0         -      10.207.29.10             -    
olive RUNNING 0         -      10.207.29.2, 10.207.39.2 -    
ubuntu@skynet3:~$ sudo lxc-stop -n oel73
ubuntu@skynet3:~$ sudo lxc-stop -n olive
ubuntu@skynet3:~$ sudo lxc-start -n oel73 -l trace -o oel73-on-ovs.txt
lxc-start: tools/lxc_start.c: main: 366 The container failed to start.
lxc-start: tools/lxc_start.c: main: 368 To get more details, run the container in foreground mode.
lxc-start: tools/lxc_start.c: main: 370 Additional information can be obtained by setting the --logfile and --logpriority options.
ubuntu@skynet3:~$ sudo lxc-start -n olive -l trace -o olive-after-oel73-is-on-ovs.txt
lxc-start: tools/lxc_start.c: main: 366 The container failed to start.
lxc-start: tools/lxc_start.c: main: 368 To get more details, run the container in foreground mode.
lxc-start: tools/lxc_start.c: main: 370 Additional information can be obtained by setting the --logfile and --logpriority options.
ubuntu@skynet3:~$ 

## and I will attach those logs also.

[oel73-on-ovs.txt](https://github.com/lxc/lxc/files/1129736/oel73-on-ovs.txt)
[olive-after-oel73-is-on-ovs.txt](https://github.com/lxc/lxc/files/1129737/olive-after-oel73-is-on-ovs.txt)

Thank you again Christian and if you need any other info please let me know.  

These configs have been working for over 2 years without issue on Ubuntu Linux typically 16.04 or 17.04 (more recently).  I tested them in the last few days on both 17.04 and 16.04 and had the same new issues which seem like the new issues could be correlated with the appearance of lxc 2.0.8 because this is a new issue for me.

**Comment by brauner** [comment]
@gstanden, thanks for the detailed info. So I think the problem is:

    lxc.mount.entry = /dev /var/lib/lxc/oel73/rootfs/dev none defaults,bind,create=dir 0 0

in the new config file for `oel73`. If you don't care why you can simply remove that line from your config and stop reading. :)

The bind-mount you're showing hasn't been needed for lxc to work for a long time now. `LXC` will automatically create those devices on startup (e.g. `/dev/console` etc.). Since you're running a privileged container establishing a bind-mount for `/dev` is not ideal since the container will have full control over all devices. Taken together with the fact that we've also really hardened our console handling code means that there's going to be some trouble. I guess that there's some valid confusion about which `/dev/pts/pmx` device is to be used since we effectively would likely overmount it.

**Comment by gstanden** [comment]
@brauner, thanks this does allow the container oel73 and olive to start/stop/attach normally.  However, it breaks the links in /dev/mapper.  I am working with LXC for production Oracle software including Oracle RDBMS single-instance & RAC.  I need the multipath links in /dev/mapper to be in the container.  Thus I have the lxc.mount for both /dev and also /dev/mapper so that the links from /dev/mapper to /dev/dm-* are valid.  When I make the change you suggest (which does fix the other problem) it breaks these links inside the container because the container can only see /dev/mapper now not /dev which it also needs for the links to be valid.  How would I address this issue in the newer paradigm you describe? Is there an alternative solution for making /dev available to the container other than the lxc.mount command?  

Also, please note I've tried the lxc-device syntax long ago ("lxc-device add -n oel73 ... ") which for reasons I do not understand provides storage that Oracle RDBMS rejects and will not recognize as usuable for Oracle ASM database storage, and so I opted long ago not to use that approach but rather used the lxc.mount of both /dev and /dev/mapper.

ubuntu@skynet3:~$ sudo lxc-attach -n oel73
bash-4.2# vi /etc/security/limits.conf
bash-4.2# su - oracle
Last login: Fri Jul  7 00:41:27 UTC 2017 from 10.207.29.1 on pts/4
[oracle@oel73 ~]$ ls -l /dev/mapper
total 0
lrwxrwxrwx 1 root root       7 Jul  7 00:36 asm_data1_00 -> ../dm-4  <-- no longer valid
lrwxrwxrwx 1 root root       7 Jul  7 00:36 asm_reco1_00 -> ../dm-3  <-- no longer valid
lrwxrwxrwx 1 root root       7 Jul  7 00:36 asm_sysd1_00 -> ../dm-2  <-- no longer valid
crw------- 1 root root 10, 236 Jul  7 00:36 control
lrwxrwxrwx 1 root root       7 Jul  7 00:36 ubuntu--vg-root -> ../dm-0
lrwxrwxrwx 1 root root       7 Jul  7 00:36 ubuntu--vg-swap_1 -> ../dm-1
[oracle@oel73 ~]$ exit
logout

**Comment by brauner** [comment]
I see the problem. Ok, we might need to think about this case for a little bit. As a workaround for now you should set:

    lxc.console = none

in both container's config files.

**Comment by brauner** [comment]
Note that if you want to mount the whole `/dev` you should also set `lxc.tty = 0` and `lxc.pts = 0`. If you don't do this then the container will overmount `/dev/pts/*` and also remove and symlinke `/dev/console`.

**Comment by gstanden** [comment]
Thanks very much. I'm grateful for the help. I'm glad to pay for support on this to canonical because it's always been a problem for Orabuntu-LXC project and one is needed.

I'm glad this convo is going on because I never liked the solution of mounting all of /dev (and /dev/mapper for that matter)  into the container anyway for all the obvious reasons and sure users of Orabuntu-LXC would balk as well at this,  and I had looked for a better alternative as this was never a good solution from a security pov to say the least.

In RHEL6 it had been possible using udev rules to create actual multipath devices (not symlink) in subdirectories of /dev but this was taken away from udev in RHEL7 and I don't know a way to do it without the udev rule that used to be available for the NAME udev attribute.

Howevern for reference here is a way that we can present disks "per container" so that containers only get the disks they are supposed to see, and no others (RHEL6)  (of course still has the problem of mounting /dev in container but at least only a subdir).

The method uses some LXC HOST UDEV rules as shown below.  

this at least makes presenting from the host more secure and allows presenting only disks to each container that belong to that container.

There will be further mods to this, but to give the general idea:

Container name is:  ol6c10

[root@oracle67 rules.d]# pwd
/etc/udev/rules.d
[root@oracle67 rules.d]# more /etc/udev/rules.d/12-dm-permissions.rules | grep grid | grep -v '#'

ENV{DM_UUID}=="mpath-23766326664353333", NAME="ol6c10/asmlib_data1_00", SYMLINK+="asmlib_devices/asmlib_data1_00", OWNER:="grid", GROUP:="asmadmin", MODE:="660"
ENV{DM_UUID}=="mpath-26537623737636234", NAME="ol6c10/asmlib_sysd1_00", SYMLINK+="asmlib_devices/asmlib_sysd1_00", OWNER:="grid", GROUP:="asmadmin", MODE:="660"

[root@oracle67 rules.d]# ls -l /dev/ol6c10/
total 0
brw-rw----. 1 grid asmadmin 252, 4 Aug  5 22:58 asmlib_data1_00
brw-rw----. 1 grid asmadmin 252, 3 Aug  5 22:58 asmlib_sysd1_00

[root@oracle67 rules.d]# cat /container/ol6c10/config | grep '/dev/ol6c10'

lxc.mount.entry = /dev/ol6c10 /container/ol6c10/rootfs/dev/ol6c10 none defaults,bind,create=dir 0 0  <-- This directive must be added to config

This way, container only sees the disks that belong to it, i.e. "ol6c10".

**Comment by gstanden** [comment]
I should add that I'll have to revisit if it's possible to use this NAME udev attribute to do the /dev subdir technique in Ubuntu LXC hosts but as I recall I could not get it to work in Ubuntu either.

**Comment by gstanden** [comment]
@brauner The workarounds worked fyi.  Thanks!  Hoping eventually for a better solution. :)

**Comment by gstanden** [comment]
@brauner some linux distro specific odd effects: Using the workaround you provided (thanks very much!) it seems Oracle Linux 7 containers are fine, but Oracle Linux 6 containers are now affected by a new problem "PTY allocation request failed on channel 0" which can be worked-around by doing an lxc-attach to the Oracle Linux 6 LXC container and running as root "mount devpts /dev/pts -t devpts" which has to be done after every reboot of server, while Oracle Linux 7 containers are NOT affected by this.
ubuntu@ubuntu:~/Downloads/orabuntu-lxc-master/orabuntu$ ssh oracle@ora73c14
The authenticity of host 'ora73c14 (xx.xxx.xx.xx)' can't be established.
ECDSA key fingerprint is SHAxxxxx.
Are you sure you want to continue connecting (yes/no)? yes
Warning: Permanently added 'ora73c14,xx.xxx.xx.xx' (ECDSA) to the list of known hosts.
oracle@ora73c14's password: 
[oracle@ora73c14 ~]$ exit
logout
Connection to ora73c14 closed.
ubuntu@ubuntu:~/Downloads/orabuntu-lxc-master/orabuntu$ ssh oracle@ora69c10
oracle@ora69c10's password: 
PTY allocation request failed on channel 0
^Cubuntu@ubuntu:~/Downloads/orabuntu-lxc-master/orabuntu$ sudo lxc-attach -n ora69c10
bash-4.1# mount devpts /dev/pts -t devpts
bash-4.1# exit
exit
ubuntu@ubuntu:~/Downloads/orabuntu-lxc-master/orabuntu$ ssh oracle@ora69c10
oracle@ora69c10's password: 
[oracle@ora69c10 ~]$

**Comment by gstanden** [comment]
@brauner In case it helps anyone, I have a tentative method for creating container-specific device nodes in /dev so that just /dev/containername subdirectory needs to be mounted into the container and no symlinks involved.  

Here are the rules that will work in RHEL7 (after NAME= was taken away between RHEL6 and RHEL7):

Example:

Here are the standard rules promulgated for RHEL7 as usual.  They can be left as-is and together with the /etc/multipath.conf file they provide the standard RHEL7 multipath storage paradigm.  The problem with this paradigm is that both /dev/mapper and /dev have to be presented to container because of the use of symlinks which isn't a good solution for containers.

ENV{DM_UUID}=="mpath-23462323936393766", SYMLINK+="asm/asm_sysd1_00", OWNER:="grid", GROUP:="asmadmin", MODE:="660"
ENV{DM_UUID}=="mpath-23766326664353333", SYMLINK+="asm/asm_data1_00", OWNER:="grid", GROUP:="asmadmin", MODE:="660"
ENV{DM_UUID}=="mpath-23930613631393537", SYMLINK+="asm/asm_reco1_00", OWNER:="grid", GROUP:="asmadmin", MODE:="660"

Here are the new rules that are used to get storage of the form:  

/dev/<container_name>/friendly_lun_name

I modified this rule from a post I saw here:  http://houseofbrick.com/udev-rules/  (see the comment by "Mike" who provided the rule basic structure, which I modified by adding a "mkdir -p" to create the subdirectory under /dev ).

Next I will try presenting the storage to container using something like:

lxc.mount.entry = /dev/ora73c10 /var/lib/lxc/ora73c10/rootfs/dev/mapper none defaults,bind,create=dir 0 0

and build Oracle RDBMS RAC on it to verify ASM likes it OK.

KERNEL=="sd?1", SUBSYSTEM=="block", PROGRAM=="/lib/udev/scsi_id -g -u -d /dev/$parent", RESULT=="23462323936393766", RUN+="/bin/sh -c '/bin/mkdir -p /dev/ora73c10; /bin/mknod /dev/ora73c10/asm_sysd1_00 b $major $minor; chown grid:asmadmin /dev/ora73c10/asm_sysd1_00; chmod 0660 /dev/ora73c10/asm_sysd1_00'"

KERNEL=="sd?1", SUBSYSTEM=="block", PROGRAM=="/lib/udev/scsi_id -g -u -d /dev/$parent", RESULT=="23766326664353333", RUN+="/bin/sh -c '/bin/mkdir -p /dev/ora73c10; /bin/mknod /dev/ora73c10/asm_data1_00 b $major $minor; chown grid:asmadmin /dev/ora73c10/asm_data1_00; chmod 0660 /dev/ora73c10/asm_sysd1_00'"

KERNEL=="sd?1", SUBSYSTEM=="block", PROGRAM=="/lib/udev/scsi_id -g -u -d /dev/$parent", RESULT=="23930613631393537", RUN+="/bin/sh -c '/bin/mkdir -p /dev/ora73c10; /bin/mknod /dev/ora73c10/asm_reco1_00 b $major $minor; chown grid:asmadmin /dev/ora73c10/asm_reco1_00; chmod 0660 /dev/ora73c10/asm_sysd1_00'"

**Comment by brauner** [comment]
We've decided to not disable mounting the whole `/dev` into a container. However, the logs will now log a `WARN()` when users do this and inform them that this is potentially dangerous as you're essentially giving the container - especially dangerous for privileged containers - access to the host's  `/dev`.
